### PR TITLE
ModularTokenizer

### PR DESCRIPTION
**Pull request by floccinauc**
Added/implemented miscellaneous methods in ModularTokenizer, and created modular_tokenizer_ops to allow antibody language model to use ModularTokenizer with minimal code changes

### CODE REVIEW

File: fusedrug/data/tokenizer/modulartokenizer/configs/tokenizer_add_special_tokens_config.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+paths:
```
Comment (YoelShoshan) [inline fusedrug/data/tokenizer/modulartokenizer/configs/tokenizer_add_special_tokens_config.yaml:1]: consider deleting this file

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 2]
```diff
@@ -333,12 +412,33 @@ def fix_json_paths(
                 )
             return loaded_conf
 
-        loaded_conf = OmegaConf.load(os.path.join(path, "config.yaml"))
-        loaded_conf_fixed = fix_json_paths(loaded_conf, path)
-        return ModularTokenizer(tokenizers_info=loaded_conf_fixed, load_adjusted_jsons=True)
+        try:
+            loaded_conf: omegaconf.dictconfig.DictConfig = OmegaConf.load(os.path.join(path, "config.yaml"))
+        except:
+            raise Exception(f"couldn't load config.yaml from {path}")
```
Comment (YoelShoshan) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:418]: please also print the exception (e.g. traceback.print_exc())

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 3]
```diff
@@ -707,20 +826,38 @@ def encode_list(
...
+                max_len = self.max_len
+
         if max_len is not None:
             merged_encoding.truncate(max_length=max_len)
-        if padding_token_id is not None:
-            # TODO: find the actual padding token ID
-            rnd_type = list(self.tokenizers_info.keys())[0]
-            rnd_inst = self.tokenizers_info[rnd_type]["tokenizer_inst"]
+
+        rnd_type = list(self.tokenizers_info.keys())[0]
```
Comment (YoelShoshan) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:836]: what does rnd_type mean ?

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 4]
```diff
@@ -10,16 +10,23 @@
 import transformers
```
Comment (YoelShoshan) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:10]: please use a consistent name - modular_tokenizer or multi_tokenizer ,
and use it for all classes and filenames
(the file here is multi_tokenizer and the class name is modular tokenizer)

File: fusedrug/data/tokenizer/modulartokenizer/configs/tokenizer_add_special_tokens_config.yaml (hunk 1 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/configs/tokenizer_add_special_tokens_config.yaml:1]: Done

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 5]
```diff
@@ -43,6 +50,16 @@ def __init__(
                 from modular jsons is best done through the load_from_jsons method. Defaults to False.
             special_tokens_dict (Optional[Dict], optional): A dictionary of special tokens that should be common among all tokenizers, with keys
                 from ["bos_token", "eos_token", "unk_token", "sep_token", "pad_token", "cls_token", "mask_token"]
+            additional_tokens_list (Optional[List], optional): A list of additional token names (str) that need to be added to the special
+                tokens of the new modular tokenizer (will be in all sub-tokenizers). Defaults to None (i.e. no tokens to be added)
+            max_possible_token_id (Optional[int], optional): An upper limit to a token ID. When IDs of tokens added to modular tokenizer
+                go above this, an exception is thrown. Defaults to None (i.e. no limit is set).
+                TODO: Currently not fully implemented. Only used during tokenizer creation.
```
[the remark this answers is not in the document]
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:57]: No longer relevant - fully implemented in all token additions, and documented a thorough description of how to add new tokens in add_special_tokens()

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py (hunk 2 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:418]: Done

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py (hunk 3 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:836]: Removed those. The point was to access any of the sub-tokenizers and call its respective token_to_id (since padding token is special, it doesn't matter which sub-tokenizer to use). I changed it to call self.token_to_id.

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 6]
```diff
@@ -594,20 +695,37 @@ def set_field(tokenizers_info_cfg: List, name: str, key: str, val: Any) -> List:
...
         with open(os.path.join(path, "config.yaml"), "w") as f:
-            OmegaConf.save(tokenizers_info_cfg, f)
+            OmegaConf.save(tokenizer_config_overall, f)
 
     def _add_single_tokenizer(
         self,
         tokenizer_info: Dict,
     ) -> None:
         raise Exception("Not implemented")
+        # self.build_inner_decoder()
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:712]: Why did you add it?

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 7]
```diff
@@ -594,20 +695,37 @@ def set_field(tokenizers_info_cfg: List, name: str, key: str, val: Any) -> List:
...
+        #     if self._get_max_mapped_id() > self._max_possible_token_id:
+        #         raise Exception(
+        #             f"tokenizer remapping resulted in IDs greater (max_id={self._get_max_mapped_id()}) than max_possible_id ({self._max_possible_token_id}). Reinitialize the modular tokenizer with larger max_possible_id"
+        #         )
 
     def add_tokenizers(
         self,
     ) -> None:
         raise Exception("Not implemented")
+        # self.build_inner_decoder()
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:723]: And this one?

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 8]
```diff
@@ -806,7 +1099,13 @@ def add_tokens(self, tokens: Union[List, str]) -> int:
         Returns:
             :obj:`int`: The number of tokens that were created in the vocabulary
         """
-        pass
+        raise Exception("Not implemented")
+        # self.build_inner_decoder()
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:1103]: And this

File: fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+        if self._verbose:
+            self._debug_max_tokenized_len_encountered = defaultdict(int)
+
+    # note: use normalizer.Sequence to chain multiple normalizers
+    def set_normalizer(self, normalizer: Any) -> None:
+        # self._tokenizer.normalizer = normalizer
+        raise Exception("Not implemented")
+
+    # note: use pre_toknizers.Sequence to chain multiple pre_toknizers
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py:90]: Are we planning to support it?
If not consider to delete them

File: fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+                f"Expected key_in={key_in} to point to a list of inputs, and instead got a {type(data_lst)}. value={data_lst}"
+            )
+
+        if self._validate_ends_with_eos is not None:
+            if not data_lst[-1].input_string.rstrip().endswith(self._validate_ends_with_eos):
+                raise Exception(
+                    f"self._validate_ends_with_eos was set to {self._validate_ends_with_eos}, but about to encode a string that does not end with it. The str end was: {data_lst[-1].input_string}"
+                )
+
+        encoded = self._tokenizer.encode_list(data_lst)
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py:187]: What are the differences from FastTokenizer op?
Isn't just this line and lines 174-179?

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py (hunk 4 above)
Comment (mosheraboh) [reply fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:10]: What he said ^

**Review by mosheraboh — APPROVED** [review]
Looks good.
See inline.

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py (hunk 4 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:10]: Done

File: fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py (hunk 10 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py:187]: That's mostly it. There are several more changes - I no longer use get_vocab(), (since it is not well-defined for modulartokenizer) which is used in several places in FastTokenizer ops (finding sentinel IDs, max token IDS, etc. - I implemented these using either get_special_vocab, which is well-defined, or built-in tokenizer functions)

File: fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py (hunk 9 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py:90]: Deleted them - will appear in the next PR
